mscore/palette.cpp

```diff
@@ -730,12 +765,306 @@ void Palette::leaveEvent(QEvent*)
       {
       if (currentIdx != -1) {
             QRect r = idxRect(currentIdx);
-            currentIdx = -1;
+            PaletteBox* pb = mscore->getPaletteBox();
+            if (!filterActive || !pb->getKeyboardNavigation())
+                  currentIdx = -1;
             update(r);
             }
       }
 
 //---------------------------------------------------------
+//   nextPaletteElement
+//---------------------------------------------------------
+void Palette::nextPaletteElement()
+      {
+      if (!filterActive)
+            return;
+      int i = currentIdx;
+      if (i == -1)
+            return;
+      i++;
+      if (i < size() && cellAt(i)) {
+            currentIdx = i;
+            }
+      update();
+      }
+
+//---------------------------------------------------------
+//   prevPaletteElement
+//---------------------------------------------------------
+
+void Palette::prevPaletteElement()
+      {
+      if (!filterActive)
+            return;
+      int i = currentIdx;
+      if (i == -1)
+            return;
+      i--;
+      if (i >= 0 && cellAt(i)) {
+            currentIdx = i;
+            }
+      update();
+      }
+
+//---------------------------------------------------------
+//   applyPaletteElement
+//---------------------------------------------------------
+
+void Palette::applyPaletteElement()
```

Comment:
Anything can be changed :-).  The point would be to figure out how much code is truly shared by these two functions and then create a helper function that does the real work.  Then the duplicated code can be removed from both applyPaletteElement and mouseDoubleClickedEvent and replaced with a call to this new helper function.  Eg, mouseDoubleClickedEvent would have those few lines as the top processing the event itself to get "i".  After that pretty much everything else should be shareable, I would think.